okio/src/test/java/okio/RealBufferedSourceTest.java

```diff
@@ -215,4 +232,181 @@
+    BufferedSource bufferedSource = new RealBufferedSource(source);
+
+    Reader reader = bufferedSource.reader();
+    String stringFromReader = readerToString(reader);
+
+    assertEquals("ab", stringFromReader);
+  }
+
+  @Test public void readerWithUtf8Bom() throws Exception {
+    Buffer source = new Buffer();
+    source.write(ByteString.decodeHex("efbbbf"));
+    source.writeUtf8("utf8");
+    BufferedSource bufferedSource = new RealBufferedSource(source);
+
+    Reader reader = bufferedSource.reader();
+    String stringFromReader = readerToString(reader);
+
+    assertEquals("utf8", stringFromReader);
+  }
+
+  @Test public void readerWithUtf16BigEndianBom() throws Exception {
+    Buffer source = new Buffer();
+    source.write(ByteString.decodeHex("feff"));
+    source.writeString("utf16", Charset.forName("UTF-16BE"));
+    BufferedSource bufferedSource = new RealBufferedSource(source);
+
+    Reader reader = bufferedSource.reader();
+    String stringFromReader = readerToString(reader);
+
+    assertEquals("utf16", stringFromReader);
+  }
+
+  @Test public void readerWithUtf16LittleEndianBom() throws Exception {
+    Buffer source = new Buffer();
+    source.write(ByteString.decodeHex("fffe"));
+    source.writeString("utf16", Charset.forName("UTF-16LE"));
+    BufferedSource bufferedSource = new RealBufferedSource(source);
+
+    Reader reader = bufferedSource.reader();
+    String stringFromReader = readerToString(reader);
+
+    assertEquals("utf16", stringFromReader);
+  }
+
+  @Test public void readerWithUtf32BigEndianBom() throws Exception {
+    Buffer source = new Buffer();
+    source.write(ByteString.decodeHex("0000feff"));
+    source.writeString("utf32", Charset.forName("UTF-32BE"));
+    BufferedSource bufferedSource = new RealBufferedSource(source);
+
+    Reader reader = bufferedSource.reader();
+    String stringFromReader = readerToString(reader);
+
+    assertEquals("utf32", stringFromReader);
+  }
+
+  @Test public void readerWithUtf32LittleEndianBom() throws Exception {
+    Buffer source = new Buffer();
+    source.write(ByteString.decodeHex("fffe0000"));
+    source.writeString("utf32", Charset.forName("UTF-32LE"));
+    BufferedSource bufferedSource = new RealBufferedSource(source);
+
+    Reader reader = bufferedSource.reader();
+    String string = readerToString(reader);
+
+    assertEquals("utf32", string);
+  }
+
+  @Test public void readerWithBomCrossingSegmentBoundary() throws Exception {
+    Buffer source = new Buffer();
+    source.writeUtf8(TestUtil.repeat('a', Segment.SIZE - 1));
+    source.write(ByteString.decodeHex("efbbbf"));
+    source.writeUtf8("utf8");
+    BufferedSource bufferedSource = new RealBufferedSource(source);
+
+    bufferedSource.readByteString(Segment.SIZE - 1);
+    Reader reader = bufferedSource.reader();
+    String stringFromReader = readerToString(reader);
+
+    assertEquals("utf8", stringFromReader);
+  }
+
+  @Test public void readerUtf8() throws Exception {
```

Comment:
You can avoid having to write each of these 2x by using BufferedSourceTest which runs each test on a variety of implementations.